// @teppeis/multimaps/dist/src/arraymultimap.js
/*!
 * melonJS Game Engine - v14.1.2
 * http://www.melonjs.org
 * melonjs is licensed under the MIT License.
 * http://www.opensource.org/licenses/mit-license
 * @copyright (C) 2011 - 2022 Olivier Biot (AltByte Pte Ltd)
 */
import { commonjsGlobal } from '../../../../../_virtual/_commonjsHelpers.js';
import { __exports as arraymultimap } from '../../../../../_virtual/arraymultimap.js';
export { __exports as default } from '../../../../../_virtual/arraymultimap.js';
import './multimap.js';
import { __exports as multimap } from '../../../../../_virtual/multimap.js';

var __extends = (commonjsGlobal && commonjsGlobal.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(arraymultimap, "__esModule", { value: true });
arraymultimap.ArrayMultimap = void 0;
var multimap_1 = multimap;
var ArrayMultimap = /** @class */ (function (_super) {
    __extends(ArrayMultimap, _super);
    function ArrayMultimap(iterable) {
        return _super.call(this, new ArrayOperator(), iterable) || this;
    }
    Object.defineProperty(ArrayMultimap.prototype, Symbol.toStringTag, {
        get: function () {
            return "ArrayMultimap";
        },
        enumerable: false,
        configurable: true
    });
    return ArrayMultimap;
}(multimap_1.Multimap));
arraymultimap.ArrayMultimap = ArrayMultimap;
var ArrayOperator = /** @class */ (function () {
    function ArrayOperator() {
    }
    ArrayOperator.prototype.create = function () {
        return [];
    };
    ArrayOperator.prototype.clone = function (collection) {
        return collection.slice();
    };
    ArrayOperator.prototype.add = function (value, collection) {
        collection.push(value);
        return true;
    };
    ArrayOperator.prototype.size = function (collection) {
        return collection.length;
    };
    ArrayOperator.prototype.delete = function (value, collection) {
        var index = collection.indexOf(value);
        if (index > -1) {
            collection.splice(index, 1);
            return true;
        }
        return false;
    };
    ArrayOperator.prototype.has = function (value, collection) {
        return collection.includes(value);
    };
    return ArrayOperator;
}());
