// @teppeis/multimaps/dist/src/setmultimap.js
/*!
 * melonJS Game Engine - v14.1.2
 * http://www.melonjs.org
 * melonjs is licensed under the MIT License.
 * http://www.opensource.org/licenses/mit-license
 * @copyright (C) 2011 - 2022 Olivier Biot (AltByte Pte Ltd)
 */
import { commonjsGlobal } from '../../../../../_virtual/_commonjsHelpers.js';
import { __exports as setmultimap } from '../../../../../_virtual/setmultimap.js';
export { __exports as default } from '../../../../../_virtual/setmultimap.js';
import './multimap.js';
import { __exports as multimap } from '../../../../../_virtual/multimap.js';

var __extends = (commonjsGlobal && commonjsGlobal.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(setmultimap, "__esModule", { value: true });
setmultimap.SetMultimap = void 0;
var multimap_1 = multimap;
var SetMultimap = /** @class */ (function (_super) {
    __extends(SetMultimap, _super);
    function SetMultimap(iterable) {
        return _super.call(this, new SetOperator(), iterable) || this;
    }
    Object.defineProperty(SetMultimap.prototype, Symbol.toStringTag, {
        get: function () {
            return "SetMultimap";
        },
        enumerable: false,
        configurable: true
    });
    return SetMultimap;
}(multimap_1.Multimap));
setmultimap.SetMultimap = SetMultimap;
var SetOperator = /** @class */ (function () {
    function SetOperator() {
    }
    SetOperator.prototype.create = function () {
        return new Set();
    };
    SetOperator.prototype.clone = function (collection) {
        return new Set(collection);
    };
    SetOperator.prototype.add = function (value, collection) {
        var prev = collection.size;
        collection.add(value);
        return prev !== collection.size;
    };
    SetOperator.prototype.size = function (collection) {
        return collection.size;
    };
    SetOperator.prototype.delete = function (value, collection) {
        return collection.delete(value);
    };
    SetOperator.prototype.has = function (value, collection) {
        return collection.has(value);
    };
    return SetOperator;
}());
